clang/lib/Parse/ParseContracts.cpp

```diff
@@ -0,0 +1,437 @@
+                                          ThisScope);
+  bool IsInvalid = false;
+  SourceLocation StartLoc = Tok.getLocation();
+
+  SmallVector<ContractStmt *, 4> Contracts;
+  while (isFunctionContractKeyword(Tok)) {
+    bool IsInvalidTmp = false;
+    StmtResult Contract =
+        ParseFunctionContractSpecifierImpl(ReturnTypeResolver, EnterScope ? CSO_ParentContext : CSO_FunctionContext, IsInvalidTmp);
+    IsInvalid |= IsInvalidTmp;
+    if (Contract.isUsable())
+      Contracts.push_back(Contract.getAs<ContractStmt>());
+  }
+  ContractSpecifierDecl *Seq = Actions.ActOnFinishContractSpecifierSequence(
+      Contracts, StartLoc, IsInvalid);
+
+  assert(DeclarationInfo.Contracts == nullptr && "Already have contracts?");
+
+  DeclarationInfo.Contracts = Seq;
+}
+
+StmtResult Parser::ParseFunctionContractSpecifierImpl(
+    llvm::function_ref<QualType()> ReturnTypeResolver, ContractScopeOffset ScopeOffset, bool &IsInvalid) {
+  assert(isAnyContractKeyword(Tok) && "Not a contract keyword?");
+  ContractKind CK = getContractKeyword(Tok).value();
+  assert((CK == ContractKind::Assert || ReturnTypeResolver) &&
+         "Missing return type resolver for function contract sequence");
+  assert((ScopeOffset == CSO_FunctionContext || CK != ContractKind::Assert) &&
+         "Incorrect scope offset for contract assert");
+  auto SetInvalidOnExit = llvm::make_scope_exit([&]() { IsInvalid = true; });
+
+  const char *CKStr = getContractKeywordStr(CK);
+
+  SourceLocation KeywordLoc = Tok.getLocation();
+  ConsumeToken();
+
+  ParsedAttributes CXX11Attrs(AttrFactory);
+  MaybeParseCXX11Attributes(CXX11Attrs);
+
+  if (Tok.isNot(tok::l_paren)) {
+    Diag(Tok, diag::err_expected_lparen_after) << CKStr;
+    SkipUntil({tok::equal, tok::l_brace, tok::arrow, tok::kw_try, tok::comma,
+               tok::l_paren},
+              StopAtSemi | StopBeforeMatch);
+
+    return StmtError();
+  }
+
+  BalancedDelimiterTracker T(*this, tok::l_paren);
+  SourceLocation ExprLoc = Tok.getLocation();
+
+  if (T.expectAndConsume(diag::err_expected_lparen_after, CKStr,
+                         tok::r_paren)) {
+    return StmtError();
+  }
+
+  ParseScope ContractScope(this, Scope::DeclScope | Scope::ContractAssertScope);
+  EnterExpressionEvaluationContext EC(
+      Actions, Sema::ExpressionEvaluationContext::PotentiallyEvaluated);
+
+  ResultNameDecl *RND = nullptr;
+  // FIXME(EricWF): We allow parsing the result name declarator in `pre` so we
+  // can diagnose it but we don't do the same for contract assert... Should we?
+  if ((CK != ContractKind::Assert) && Tok.is(tok::identifier) &&
+      NextToken().is(tok::colon)) {
+    // Let this parse for non-post contracts. We'll diagnose it later.
+
+    IdentifierInfo *Id = Tok.getIdentifierInfo();
+    SourceLocation IdLoc = ConsumeToken();
+
+    ExprLoc = ConsumeToken();
+    QualType ReturnType;
+    if (ReturnTypeResolver) {
+      ReturnType = ReturnTypeResolver();
+    }
```

Comment:
```suggestion
    if (ReturnTypeResolver)
      ReturnType = ReturnTypeResolver();
```

just because i noticed :)